Added working support for private storage

nickovs opened the pull request:
## Description:

This is a fixed version of #16878. It adds support for marking `Store` objects objects as private to prevent data exposure.

Compared to the original PR it:
* fixes the corruption bug in the original PR caused by not properly truncating the storage file when writing JSON files using `utils.json.save_json()`
* Implements atomic file replacement in the `helpers.storage.Store` class to prevent file corruption if the system crashes while writing a storage file.
* Ensures that the file permissions of existing storage files get updated if they are now marked as private and the contents is updated.
* Adds test cases to `test/helpers/test_storage.py` to test that updating an existing storage file works correctly.

**Related issue (if applicable):** fixes #16349

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** *N/A*

## Example entry for `configuration.yaml` (if applicable):
*N/A*

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. 

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) *N/A*

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]). *N/A*
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]). *N/A*
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`. *N/A*
  - [x] New files were added to `.coveragerc`. *N/A*

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works. *N/A*

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

nickovs committed:
Fixed file corruption bugs in private storage code.

nickovs committed:
Merge branch 'dev' into private-storage-2

nickovs commented:
@amelchio I fixed the merge conflict but it seems that you also reverted a fix to a broken test file (`tests/components/emulated_hue/test_init.py`) and I don't know how to put it back without pushing a whole new pull request. Any suggestions?

nickovs committed:
Restoring fixed test case.

nickovs commented:
OK, I just added the file back at my end and pushed it again. It seems to work now.

pvizeli commented on homeassistant/helpers/storage.py:
```diff
@@ -187,7 +187,9 @@ def _write_data(self, path: str, data: Dict):
 
         _LOGGER.debug('Writing data for %s', self.key)
-        json.save_json(path, data, self._private)
+        tmp_path = path + "__TEMP__"
```
Mark file as not visible with `.`, and we should use a context manager to cleanup the temporary file after this.

balloob replied:
No need to clean up since we're using os.replace

amelchio commented:
@nickovs Yes, fine to handle the revert like that 👍. You can push additional commits until the PR is merged, that will also work to address review comments.

balloob commented on tests/helpers/test_storage.py:
```diff
@@ -30,6 +31,14 @@ def store(hass):
     assert data == MOCK_DATA
 
 
+async def test_overwriting(hass, store):
```
This actually doesn't do anything because all writes for the storage helper are mocked in tests

balloob replied:
Correct, we try to avoid I/O as much as possible in unit tests as it's slow.

We should still test this functionality but it should test `util.save_json` directly, as that one is not mocked out.

nickovs committed:
Implemented test suite for utils/json.py

nickovs committed:
Added new unit test cases for util/json.py

nickovs committed:
Dixed formatting nags

nickovs committed:
Fixed more nags from the Hound

nickovs committed:
Added doc strings to some very short functions

nickovs committed:
Fixing lint's complains about my choice of parts of speach. Sigh.

nickovs commented:
OK. I wrote a test suite for the underlying `util/json.py` code, which was largely untested before. These tests of necessity do some I/O, since they are testing file system operations, but they take very little time to do so and it's good to have this code tested. There are still some pathways between the upper layers of the storage system and the low-level file representation but, as discussed above in the code comments, this would require some rather different test suites to be written.

balloob commented on homeassistant/helpers/storage.py:
```diff
@@ -187,7 +187,9 @@ def _write_data(self, path: str, data: Dict):
-        json.save_json(path, data, self._private)
+        tmp_path = path + "__TEMP__"
+        json.save_json(tmp_path, data, self._private)
+        os.replace(tmp_path, path)
```
I think that we should move this functionality to `save_json` so that all calls can benefit. 

I also think that we should consider cleaning up the temp path in case of an exception replacing the other file

nickovs replied:
Fair points. I'll make it so.

nickovs committed:
Moved atomic save operations down into util/json.py so that all benefit.

Added extra clean-up code to ensure that temporary files are removed in
case of errors.
Updated emulated_hue unit tests to avoid errors.

nickovs committed:
Apparently 'e' is not allows as a variable name for an exception...

balloob commented:
Alright let's try again, great work! 👍
